Bind PIDs (persistent identifiers) to files, directories and AIPs

jrwdunham opened the pull request:
This feature allows Archivematica to be configured to assign persistent identifiers (PIDs) to files, directories and AIPs. The PIDs are documented in the AIP/DIP METS and are configured to resolve to user-configured URLs. Auxiliary features:

- Handle server configuration in dashboard
- "Bind PIDs" and "Assign UUIDs to directories" configurable choices
- PID binding works with arranged files and directories, and with extraction
- Accession number from transfer can be used as AIP/DIP PID (instead of default UUID)
- Empty directories in objects/ are now documented in METS file and contain empty .keep files to prevent deletion by BagIt

sevein commented:
@jrwdunham, I've rebased your branch because the build was failing for a different reason (see #693).

jrwdunham committed:
PID Binding

This feature allows Archivematica to be configured to assign persistent
identifiers (PIDs) to files, directories and AIPs. The PIDs are documented in
the AIP/DIP METS and are configured to resolve to user-configured URLs.
Auxiliary features:

- Handle server configuration in dashboard
- "Bind PIDs" and "Assign UUIDs to directories" configurable choices
- PID binding works with arranged files and directories, and with extraction
- Accession number from transfer can be used as AIP/DIP PID (instead of default
  UUID)
- Empty directories in objects/ are now documented in METS file and contain
  empty .keep files to prevent deletion by BagIt

jrwdunham commented:
See https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/15 to run the applicable acceptance tests (Gherkin features)

sevein commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -1039,17 +1126,25 @@ def write_mets(tree, filename):
+    parser.add_option("-x", "--xmlFile", action="store", dest="xmlFile",
+                      default="")
+    parser.add_option("-a", "--amdSec", action="store_true", dest="amdSec",
+                      default=False)
```
This block is only for cosmetic changes?

jrwdunham replied:
Yeah. Sorry for the noise.

sevein commented on src/dashboard/src/components/filesystem_ajax/views.py:
```diff
@@ -333,11 +333,30 @@ def copy_to_start_transfer(filepath='', type='', accession='', transfer_metadata
+
 def create_arranged_sip(staging_sip_path, files, sip_uuid):
-    shared_dir = django_settings.SHARED_DIRECTORY
+    shared_dir = helpers.get_server_config_value('sharedDirectory')
```
I think this change should be deleted, `get_server_config_value` was recently deleted.

jrwdunham replied:
Good catch. Should be fixed by 06b38824053cc7fa89e60e37490d8a0d35630113.

jrwdunham reviewed (commented):
The needed mods to the default and automatic processing config constants have been added. See 5149715f6ff40e17e71d63e1ccbc85a0559b4a11.

jrwdunham committed:
Getting shared directory form Django settings

jrwdunham committed:
Bind PIDs choices in processing config constants

In both the default and automatic processing configs we have:

- Bind PIDs? No.
- Assign UUIDs to directories? No.

sevein commented:
Rebased.

sevein commented on src/MCPClient/lib/clientScripts/assignUUIDsToDirectories.py:
```diff
@@ -0,0 +1,141 @@
+    being the root directory of the transfer with UUID ``transfer_uuid``. Do
+    this only if ``include_dirs`` is ``True``.
+    """
+    _exit_if_not_include_dirs(include_dirs)
```
As a first-timer here I didn't find this function particularly useful. Its name describes the behaviour that the caller expects only with the help of a function decorator. Is this idiomatic in Python?

jrwdunham replied:
I'm not sure I agree that the name doesn't describe the behaviour: it clearly states that the function should exit (i.e., return) under a certain condition. Just because Python doesn't succinctly let you return from function `f` by calling function `g` from `f`, doesn't mean you shouldn't be able to get around that. Note that the calls to `_get_subdir_paths` and `_get_transfer_mdl` are doing the same thing: they will cause `main` to exit if a known exception is raised too.

Would it help if I renamed `_exit_if_not_include_dirs` to `_return_if_not_include_dirs` or `_assert_include_dirs`? Or do you really just want to see `return` statements in `main`?

sevein replied:
> Note that the calls to _get_subdir_paths and _get_transfer_mdl are doing the same thing: they will cause main to exit if a known exception is raised too.

But you didn't name it as `_exit_or_get_subidr_paths`...

I think it's ok the way it is, clear enough. I think this comes down to the fact that I don't find error handling via exceptions better because the flow seems to be all over the place. This is probably me biased by having spent time using Go where you have to deal with the errors right away.

Thanks for the explanation, please leave it as it is!

sevein commented on src/MCPClient/lib/clientScripts/assignUUIDsToDirectories.py:
```diff
@@ -0,0 +1,141 @@
+            os.walk(root_path) if dir_path not in exclude_paths)
+
+
+def str2bool(val):
```
You've also defined this function in `archivematicaFunctions`.

jrwdunham replied:
And in bind_pid.py too! I like to practice RY. Just kidding: will fix.

sevein replied:
:dancing_men:

sevein commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -356,3 +358,17 @@ def deUnicode(str):
     return unicode(str).encode('utf-8')
+
+
+def create_directory_models(dir_paths_uuids, unit_mdl, unit_type='transfer'):
```
Nitpick: you're not creating models, you're creating `Directory` objects I would think.

jrwdunham replied:
I often call "model class instances" just "models" as a shorthand; given the context, I think it makes sense. Similarly, I informally type-annotate such instances with `_mdl`, e.g., `file_mdl = File()` if I think it improves readability. But I guess you're right that in this case it kinda sounds like I'm dynamically create Django classes, so I'll change it.

jrwdunham replied:
TODO: make this a method on the `Directory` class itself (thanks sevein).

sevein replied:
> I often call "model class instances" just "models" as a shorthand; given the context, I think it makes sense. Similarly, I informally type-annotate such instances with _mdl, e.g., file_mdl = File() if I think it improves readability. But I guess you're right that in this case it kinda sounds like I'm dynamically create Django classes, so I'll change it.

Makes total sense, I like it! 👍

sevein commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -0,0 +1,459 @@
+    return config
+
+
+def get_command_line_params():
```
Really nice!

sevein commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -0,0 +1,459 @@
+    response = requests.post(
+        kwargs['pid_web_service_endpoint'],
+        data=request_body,
+        headers={'Content-Type': 'text/xml',  # <= non-paramed assumpton alert
```
What does the comment mean?

jrwdunham replied:
I guess I was trying to highlight which parts of the "bind PID" request are baked-in and which are user-configurable. The fact that we have XML in the request body, the fact that we're making a POST request, etc.—these are currently non-configurable. Maybe I'll just remove it.

sevein commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -0,0 +1,459 @@
+        headers={'Content-Type': 'text/xml',  # <= non-paramed assumpton alert
+                 'Authorization': 'bearer {}'.format(
+                     kwargs['pid_web_service_key'])},
+        verify=False  # TODO: REMOVE IN PRODUCTION!!! Needed in dev because test handle endpoint has bad HTTPS
```
Why don't you make this configurable?

jrwdunham replied:
Good idea. Will do.

sevein commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -0,0 +1,459 @@
+    http://<RESOLVER_URL>/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858?locatt=view:original => https://my-domain.org/original/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858
+    http://<RESOLVER_URL>/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858?locatt=view:preservation => https://my-domain.org/preservation/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858
+    http://<RESOLVER_URL>/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858 => https://my-domain.org/access/12345/58fe1b06-ee88-41ca-9bc9-320dcac6c858
+    Joels-MacBook-Pro-2:handlepid joeldunham
```
What does this line mean?

jrwdunham replied:
It means I copy-pasted without editing :) (you mean just `Joels-MacBook-Pro-2:handlepid joeldunham`, right?)

sevein commented on src/archivematicaCommon/tests/test_bindpid.py:
```diff
@@ -0,0 +1,131 @@
+        """Set some stuff up...
+        """
+
+    def tearDown(self):
```
If unused, can we delete them?

jrwdunham replied:
Sure. Will do.

sevein commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -351,6 +352,23 @@ def processing(request):
+    if request.method == 'POST':
+        form = HandleForm(request.POST)
+        if form.is_valid():
+            models.DashboardSetting.objects.set_dict(
```
Happy to see that this is being used!

sevein commented on src/dashboard/src/components/administration/views.py:
```diff
@@ -351,6 +352,23 @@ def processing(request):
+    else:
+        form = HandleForm(
+            initial=models.DashboardSetting.objects.get_dict('handle'))
+    return render(request, 'administration/handle_config.html', locals())
```
Should we replace `locals()` with an explicit dict of things we're passing to the template?

jrwdunham replied:
I guess so. Using `locals()` like this is new to me, but it seems to be used all over the place in AM/SS when rendering templates. I thought I was just being consistent...

helenst replied:
I agree it's preferable to name what we are passing to the template (explicit is better than implicit) - it does make it easier to see what's going on and provides a more clear interface between the view and template. Perhaps we could adopt this more going forwards.

jrwdunham replied:
All right, all right, 2 against 1 and a quote from Zen: will change.

sevein commented on src/dashboard/src/templates/administration/sidebar.html:
```diff
@@ -48,7 +49,11 @@
 
-    <li class="{% active request url_language %}"><a href="{{ url_language }}">{% trans "Language" %}</a></li>
+    <li class="{% active request handle_config %}"
+        ><a href="{{ handle_config }}">{% trans "Handle server config" %}</a></li>
```
This `>` in a new line is pretty new to me, never seen this before in the Dashboard. Why are you doing it this way? Do we want to start forcing a 79col rule in HTML templates too from now on?

jrwdunham replied:
Haha. Sorry, that's just me. Everything I write has a 79col rule! (emails, notes, html) I know it's crazy. I have line-wrapping turned off in my editor, which is why you see this kind of stuff.

But there's another problem here, no? It looks like I removed the "Language" link!

sevein reviewed (changes requested):
Thanks, @jrwdunham. Looking great. I've shared some comments.

jrwdunham committed:
Bind PIDs verify certs configurable

- In Handle configuration (for PID binding) the user can now choose
  whether to verify SSL certificates when making PID binding (creation)
  requests.
- Other relatively minor changes in response to code review, including
  making multiple ``Directory`` creation a class method on the
  ``Directory`` model class.

jhsimpson reviewed (approved):
I tested this branch using the acceptance-tests https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/15 and the sampledata https://github.com/artefactual/archivematica-sampledata/pull/4

and the feature works as expected.  I think this is ready to merge.

sevein commented on src/dashboard/src/main/models.py:
```diff
@@ -414,17 +439,69 @@ class File(models.Model):
+        that the directories are a part of.
+        """
+        unit_type = {'transfer': 'transfer'}.get(unit_type, 'sip')
+        cls.objects.bulk_create([
```
Should this return? I'm thinking at some point this could be chained.

jrwdunham replied:
made it so.

sevein commented on src/dashboard/src/templates/administration/handle_config.html:
```diff
@@ -0,0 +1,35 @@
+</div>
+
+{% endblock %}
+
```
Can you delete this empty line?

jrwdunham replied:
sure.

sevein reviewed (approved):
👍  tested using am.git!

helenst commented on src/MCPClient/lib/clientScripts/assignUUIDsToDirectories.py:
```diff
@@ -0,0 +1,133 @@
+    exclude_paths = (root_path, os.path.join(root_path, 'objects'))
+    return (format_subdir_path(dir_path, root_path) for dir_path, __, ___ in
+            os.walk(root_path) if dir_path not in exclude_paths)
+
```
I find this formatting of the generator expression slightly hard to read, and beginning a new line with the `for` or `if` clause makes things clearer.

```python
return (format_subdir_path(dir_path, root_path) 
        for dir_path, __, ___ in os.walk(root_path)
        if dir_path not in exclude_paths)
```

jrwdunham replied:
yeah, that is more readable.

helenst commented on src/MCPClient/lib/clientScripts/bind_pid.py:
```diff
@@ -0,0 +1,156 @@
+    http://<PID-RESOLVER>/<NAME_AUTH>/<PID>?locatt=view:original
+        => http://my-org-domain.org/original/files/<PID>
+
+The sole command-line argument is the File's UUID. If the --pind-pids option
```
typo here (`--pind-pids`) and in a couple of other places

jrwdunham replied:
Thanks. Found "pind-pids" here and in bind_pids.py too.

helenst commented on src/archivematicaCommon/lib/bindpid.py:
```diff
@@ -0,0 +1,459 @@
+    with open(cf) as filei:
+        config.read_file(filei)
+    return {key: config.get('Handle', key, fallback=None) for key in CFGABLE_PARAMS}
+
```
For Python 2, `read_file` needs to be `readfp` and the `fallback` kwarg isn't supported

jrwdunham replied:
Good catch. On my dev machine I must have installed `configparser` in my 2.7.6 Python so I wasn't seeing this error. With the changes coming in, I have tested command-line usage of bindpid.py and it seems to be working on 2.7.6, 2.7.8 and 3.4.2.

helenst commented on src/dashboard/src/components/filesystem_ajax/views.py:
```diff
@@ -333,11 +333,30 @@ def copy_to_start_transfer(filepath='', type='', accession='', transfer_metadata
+        models.Transfer.objects.filter(
+            file__uuid__in=file_uuids).values_list(
+                'diruuids').distinct()])
+
```
If I am understanding this function right, it could be simplified to remove the need of extra Python code after querying:

e.g.
```python
return Transfer.objects.filter(file__uuid__in=file_uuids, diruuids=True).exists()
```

jrwdunham replied:
Right. That looks better. Thanks.

helenst commented on src/dashboard/src/main/models.py:
```diff
@@ -396,11 +399,33 @@ def __str__(self):
+
+    def __str__(self):
+        return (u'Identifier {i.identifiervalue} of type'
+                ' {i.identifiertype}'.format(a=self))
```
should be `i=self`

jrwdunham replied:
yup. nice catch.

helenst commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -1077,26 +1172,69 @@ def write_mets(tree, filename):
+        aipDmdSec = getDirDmdSec(sip_mdl, sip_dir_name)
+        globalDmdSecCounter += 1
+        dmdSecs.append(aipDmdSec)
+        aip_dmd_id = "dmdSec_" + globalDmdSecCounter.__str__()
```
instead of calling `__str__` directly you could do `str(globalDmdSecCounter)` which is more idiomatic

jrwdunham replied:
True. As with the `locals()` issue, this was a result of being consistent with the rest of the codebase, even if it is non-idiomatic, just like when I use camelCase to match the style of legacy code, etc. But I'll change it.

helenst reviewed (commented):
Hi @jrwdunham this is awesome! I got the acceptance tests to pass (eventually, since it was the first time I've run them) and played around with the bindpid script. I commented on a couple of small things that were causing errors, plus some style points.

jrwdunham committed:
Bind PIDs CLI works with python 2 and 3

Handles differences between Python2's ConfigParser and Python3's
configparser. This commit also includes some additional minor bugfixes and style
improvements from code review.

jrwdunham commented:
Thanks @sevein, @helenst and @jhsimpson for all the great code review; I think it improved this PR quite a bit.
